### PR TITLE
Allow for a sender retry cap

### LINKED ISSUE #216: Retry mechanism is not configurable
There is no way to configure the behavior of retries when sending a request; it will essentially retry until either the sending succeeds or the context is cancelled.
The cited problem with this approach is that callers are oblivious to scaling issues, i.e. their EH is too small for their workload and we keep swallowing "server busy" errors so they are unaware unless they manually look at metrics in the portal (perhaps there's a programmatic way to achieve this).
It could be as simple as making a retry cap available to callers.  Another option (probably the preferred solution) would be to provide a way for callers to supply their own retry policy so they have complete control over how errors are handled.

### PR DESCRIPTION
**Pull request by richardpark-msft**
As part of the work @jhendrixMSFT added for recovering sender links we also wanted to allow the customer to cap the # of retries (currently it's infinite so long as the errors are considered retryable).

This PR allows that by exposing a new option as part of the event hub client that allows you to specify the max retry count. It's a small step to also allow configuring the underlying backoff policy details as well but I've purposefully not done that so we can take some time to discuss a possible design for retry policies.

Fixes #216

### CODE REVIEW

File: hub.go [hunk 1]
```diff
@@ -738,6 +739,18 @@ func HubWithWebSocketConnection() HubOption {
 	}
 }
 
+// HubWithSenderMaxRetryCount configures the Hub to only retry sending messages `maxRetryCount` times.
```
Comment (jhendrixMSFT) [inline hub.go:742]: I would add another comment something like.
```go
// Specify a value less than 1 to disable retries.
```

File: hub.go [hunk 2]
```diff
@@ -738,6 +739,18 @@ func HubWithWebSocketConnection() HubOption {
...
+func HubWithSenderMaxRetryCount(maxRetryCount int) HubOption {
+	return func(h *Hub) error {
+		if h.senderRetryOptions == nil {
+			h.senderRetryOptions = newSenderRetryOptions()
```
Comment (jhendrixMSFT) [inline hub.go:746]: Could this be moved into the constructor for `Hub`?  Then you don't need to call it in more than one place.

File: sender.go [hunk 3]
```diff
@@ -52,6 +52,7 @@ type (
 		partitionID     *string
 		Name            string
 		recoveryBackoff *backoff.Backoff
+		maxRetries      int
```
Comment (jhendrixMSFT) [inline sender.go:55]: Should this an the above be replaced with a `senderRetryOptions` like in `Hub`?

File: sender.go [hunk 4]
```diff
@@ -64,29 +65,57 @@ type (
...
+
+func newSenderRetryOptions() *senderRetryOptions {
+	return &senderRetryOptions{
 		recoveryBackoff: &backoff.Backoff{
 			Min:    10 * time.Millisecond,
 			Max:    4 * time.Second,
 			Jitter: true,
 		},
-		cond: sync.NewCond(&sync.Mutex{}),
+		maxRetries: 5,
```
Comment (jhendrixMSFT) [inline sender.go:93]: I don't think we can do this in track 1 (at least for now) as it's a breaking change behavior-wise.

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]
I think for now we should keep the infinite retries in case people rely on this.

File: sender.go (hunk 4 above)
Comment (richardpark-msft) [reply sender.go:93]: You're absolutely right. I'll default this to infinite (and also make sure we respect infinite!)

File: hub.go (hunk 2 above)
Comment (richardpark-msft) [reply hub.go:746]: There are two spots where a Hub is constructed, so I think this could work. I'll shift this code over.

File: hub.go (hunk 1 above)
Comment (richardpark-msft) [reply hub.go]: Done.

File: hub.go (hunk 2 above)
Comment (richardpark-msft) [reply hub.go]: Done.

File: sender.go (hunk 3 above)
Comment (richardpark-msft) [reply sender.go]: Done.

File: sender.go (hunk 4 above)
Comment (richardpark-msft) [reply sender.go]: Done. Infinite is now the default.

File: sender.go [hunk 5]
```diff
@@ -64,29 +65,60 @@ type (
...
+		maxRetries: -1, // default to infinite retries
+	}
+}
+
+// newSender creates a new Service Bus message sender given an AMQP client and entity path
+func (h *Hub) newSender(ctx context.Context, retryOptions *senderRetryOptions) (*sender, error) {
+	span, ctx := h.startSpanFromContext(ctx, "eh.sender.newSender")
+	defer span.End()
+
+	if retryOptions == nil {
```
Comment (jhendrixMSFT) [inline sender.go:106]: Minor issue, and I know it sounds weird, but the idiom in Go is to _not_ check this for `nil` and let it panic.

Comment (richardpark-msft) [reply sender.go]: Done. I'm here for the idioms :)

**Review by jhendrixMSFT — APPROVED** [review]

**Comment by richardpark-msft** [comment]
(looks like the races_test.go was missing an import)

**Pull request merged**